#11 기본 에러 입셉션 핸들러 추가

seokhwan-an opened the pull request:
앞으로 개발을 진행하는데 있어서 커스텀 에러를 처리하기 위한 기본 에러 입셉선 핸들러를 추가했습니다.
### 추가적으로 한 일
- rest-assured 의존성 추가
- 인수 테스트를 하기 위한 test내 support 패키지 추가

저는 통합테스트를 진행하는데에 MockMVC보다 BDD의 성향을 띄는 rest-assured를 더 선호해서 이를 이번 작업에서 추가해 이용했습니다. 

### ErrorCode를 enum이 아닌 interface로 한 이유
ErrorCode를 enum으로 하게 되면 모든 Error 정보를 하나의 파일에 관리해야하는데 이는 여러 도메인에서 커스텀 에러가 많아질 수록 하나의 파일이 무거워진다고 생각하여 interface로 구성을 했습니다. 따라서 각 도메인 내에서 해당 errorCode를 관리할 수 있습니다.  

### exception log
<img width="1119" alt="image" src="https://github.com/choorai/excuse-me-backend/assets/70303795/d63c1be3-8d85-4c57-a5cf-133f97e094f1">
exception log는 위의 사진처럼 나타납니다!

혹시 다른 의견 있으면 편하게 리뷰 달아주세요
### 관련된 이슈
- close #11

seokhwan-an committed:
#11 feat : ErrorCode(예외 정보를 가지는 enum) 추가

seokhwan-an committed:
#11 feat : CustomException 추가

seokhwan-an committed:
#11 feat : rest-assured 의존성 추가

seokhwan-an committed:
#11 feat : GlobalExceptionHandler 추가

세부사항
- springBoot 테스트를 위한 기본 세팅 추가

seokhwan-an committed:
#11 refactor: ErrorCode enum에서 interface로 변경

seokhwan-an committed:
#11 refactor: 파라미터에 final 추가 및 불필요한 공백제거

xi-jjun commented on src/main/java/choorai/excuseme/global/exception/CustomException.java:
```diff
@@ -0,0 +1,43 @@
+    private static final String VALUE_DELIMITER = " / ";
+
+    private final ErrorCode errorCode;
+    private final Map<String, Object> inputValuesByProperty;
```
~~약간 이름이 긴 듯 한데, 추가적인 정보를 받는 역할이니깐 `additionalInfos` 정도로 변경하는 것은 어떨까요~~
아래 코드 보니깐, 에러가 발생했을 때의 입력정보들을 담기 위한 용도인 것 같은데요!
사실 에러가 발생했을 때에는 입력 정보 뿐만 아니라 다른 부가적인 정보들도 담아야할 경우가 생기는 것 같아요. 따라서 해당 컬럼의 이름을 `additionalinfos`로 저희 회사 팀에서 사용하고 있었어서 말씀드린 부분입니다!

seokhwan-an replied:
저는 추가로 나타내주는 정보는 요청에서 잘못된 정보를 나타내는 것이라고 생각해서 값의 형태(인적 정보가 요청으로 넘어오면 `이름`, `나이` 같은 정보)와 넘어온 값으로만 한정 지어서 위와 같이 네이밍을 했는데 리뷰주신 것처럼 더 포괄적인 의미를 가지는 네이밍이 이 상황에서는 더 좋은 것 같습니다!👍

xi-jjun commented on src/main/java/choorai/excuseme/global/exception/CustomException.java:
```diff
@@ -0,0 +1,43 @@
+
+@AllArgsConstructor
+@Getter
+public class CustomException extends RuntimeException {
```
공용 익셉션에서 추후 BadRequestException 등등으로의 계층이 추가될 때를 고려하면, '공통'이라는 의미의 `CommonException` 네이밍은 어떨까요?

seokhwan-an replied:
넵 이름 수정하겠습니다!👍👍

xi-jjun commented on src/main/java/choorai/excuseme/global/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    private static final String ERROR_LOG_FORMAT = "url= {}, errorCode= {}, errorMessage= {}";
```
`=` 다음에 공백이 있는 이유가 뭔지 알 수 있을까요?

seokhwan-an replied:
공백없이 붙여 놓으면 너무 바로바로 붙어 있는 느낌이라 하나의 공백을 두고 넣었는데 공백을 제거하는 것이 좋을까요?

xi-jjun replied:
저는 나중에... 혹시라도 로그 파이프라인을 만들어서 모니터링 툴에서 해당 로그를 확인할 수 있게 한다고 했을 때,
파싱이 쉽도록 붙여놓는게 좋지 않을까....? 라는 막연한 생각을 했었습니다.. (파싱에 일관성 있도록 하기 위함)

xi-jjun commented on src/test/java/choorai/excuseme/global/exception/ExceptionTestController.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Map;
+
+@RestController
+public class ExceptionTestController {
+
+    @GetMapping("/errorTest")
+    public void test() {
+        throw new CustomException(TestErrorCode.TEST_ERROR, Map.of("에러", "테스트"));
+    }
+}
```
혹시 해당 코드는 무엇을 위해 존재하는 것일까요?

seokhwan-an replied:
테스트를 하기 위해 생성한 객체입니다. 
제가 테스트를 했던 이유는 다음과 같습니다. 이번 작업 사항 중에 controller advice를 추가해야 했고 내부적으로 다음과 같은 코드가 있었습니다.
```java
@ExceptionHandler(CustomException.class)
public ResponseEntity<CustomExceptionResponse> handleException(
        final CustomException exception,
        final HttpServletRequest request
) {
    trackLog(exception, request);
    return CustomExceptionResponse.toResponseEntity(exception.getErrorCode());
}
```
이는 예외가 발생한 경우 로그를 찍고 예외정보를 json 형태로 반환해줍니다.
저는 이 부분에서 제가 의도한 json 형태로 예외사항이 반환되는지를 확실하게 하는 것이 중요하다고 생각을 해서 테스트 객체를 만들고 테스트를 진행했습니다.

xi-jjun replied:
제가 생각했던 것은 향후 에러를 만들어내는 컨트롤러가 생길거라서 그 때 테스트 케이스를 추가하면 자연스레 테스트가 될 것이라 생각했었습니다.
하지만 말씀주신 의도 파악했고, 이렇게 면밀하게 테스트를 가져가는 것 또한 테스트 커버리지가 높아져서 안정성이 높아질 것이기에 어느정도 긍정적입니다 ㅎㅎㅎ

xi-jjun reviewed (commented):
`ExceptionTestController`와 같이 현재 구현이 되지 않은 부분에 대해서 테스트를 진행하려고 억지로 만든 느낌이 조금 들어서요...
지금 현재 Exception 클래스를 만드는 것까지가 범위인데 조금 많이 벗어난 부분들이 보여서 말씀드립니다.

그리고 BDD 스타일의 RestAssured 를 사용하셨다고 말씀 주셨는데, 본인만 익숙한 부분일 수 있어서 각 코드 간략하게 설명을 추가해주시면 감사하겠습니다!

seokhwan-an committed:
#11 style: 객체 및 변수 네이밍 수정

세부사항
- 객체 : CustomException -> CommonException으로 변경
- 변수 : inputValuesByProperty -> additionalInfos로 변경

seokhwan-an committed:
#11 style: 불필요한 공백 제거

seokhwan-an commented:
> `ExceptionTestController`와 같이 현재 구현이 되지 않은 부분에 대해서 테스트를 진행하려고 억지로 만든 느낌이 조금 들어서요... 지금 현재 Exception 클래스를 만드는 것까지가 범위인데 조금 많이 벗어난 부분들이 보여서 말씀드립니다.
> 
> 그리고 BDD 스타일의 RestAssured 를 사용하셨다고 말씀 주셨는데, 본인만 익숙한 부분일 수 있어서 각 코드 간략하게 설명을 추가해주시면 감사하겠습니다!

이 부분에 대한 답을 하겠습니다. 제가 controller에 대해 테스트를 하는 방법에 대해 고민하고 학습할 때 읽었던 글입니다. ([rest-assured VS mockMvc](https://tecoble.techcourse.co.kr/post/2020-08-19-rest-assured-vs-mock-mvc/)) 이 글을 간단하게 요약을 한다면 
MockMvc는 presentLayer 계층의 bean들만 불러와서 테스트가 진행되기 때문에 controller를 단위 테스트를 하는데 주로 이용이 되고 rest- assurd의 경우에는 실제 spring boot가 돌아가는 것과 같이 실제 port를 할당 받아서 모든 bean들을 불러와서 테스트를 하는 것입니다.
```java
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class AcceptanceTest {

    @LocalServerPort
    private int port;

    // 테스트마다 랜덤포트를 할당받는다. 랜덤포트를 할당받는 이유는 테스트간의 충돌을 방지하기 위함입니다
    @BeforeEach
    void setUp() {
        RestAssured.port = port;
    }
}
```
 따라서 테스트가 동작하는 속도는 당연하게 MockMvc가 rest-assured보다 빠릅니다.

이번 코드는 단순하게 test용 controller를 만들고 거기에 동작하는 것을 테스트 하는 것이니 mockMvc를 이용하는 것이 더 속도도 빠르고 좋은 방법이라고 생각하실 수 있을 것 같습니다.
제가 이번 controller를 테스트 하는 과정에서 rest-assuerd를 이용한 또 다른 이유는 BDD(Behavior Driven Develop)형태의 코드이기 때문인데요 쉽게 설명하면 저희가 테스트를 할 때 given when then을 하면서 테스트를 명확하게 하듯이 controller에 요청을 보내는 과정이 가독성이 좋습니다.
```java
@DisplayName("예외 발생시 예외에 알맞은 응답 status와 response응답 값을 보낸다.")
@Test
void test_exception_result() {
    // given
    // when
    final CustomExceptionResponse result = RestAssured
            .when().get("/errorTest")
            .then().statusCode(HttpStatus.BAD_REQUEST.value())
            .extract()
            .body().as(CustomExceptionResponse.class);

    // then
    final CustomExceptionResponse expect = new CustomExceptionResponse(TestErrorCode.TEST_ERROR.getCode(), TestErrorCode.TEST_ERROR.getMessage());
    assertThat(result).usingRecursiveComparison()
            .isEqualTo(expect);
    }
```
위의 코드를 보면 RestAssured에 요청을 보내는 과정을 보면 given, when, then으로 어떻게 요청이 보내지는지 명확하게 나타나 있습니다. (물론 현재 `/errorTest`에 요청을 보낼 때 cookie나 requestBody가 존재하지 않아서 given이 존재하지 않는 것입니다.)
그래서 테스트의 목적과 가독성이 높다고 생각해서 rest-assured를 이용해서 테스트를 진행했습니다. 

그렇다고 제가 mockMvc를 이용하지 않는 것은 아닙니다. mockMvc를 이용하는 경우는 topdown 방식(controller -> service -> domain 순서로 개발)으로 개발이 진행되는 경우에는 controller의 단위테스트가 필요한 경우가 있어서 속도가 빠른 mockMvc를 주로 이용합니다.

JoungMinJu reviewed (approved):
리뷰 읽으면서 많이 배웠습니다! 어떤 것을 왜 적용했는지 명확히 명시되어있어서 더 좋았던 것 같아요 :)

저는 추가적인 의견은 없어서 approve하겠습니다!!

xi-jjun reviewed (approved):
approve 드리는데, 앞으로는 석주형,민주 approve만 돼도 merge 해주셔도 됩니다 ㅎㅎ

설명과 함께 꼼꼼한 테스트 케이스 작성까지 감사합니다.
수고하셨습니다.

저는 개인적으로 finish feature로 머지하기 전에 develop기준으로 현재 feature 브랜치를 rebase한 뒤, 
rebase된 feature브랜치를 remote에 force push 한 다음,
로컬에서 feature finish를 하는 편인데요 이와 같이 하면 아래와 같은 장점이 있습니다.
- git conflict가 발생할 확률이 현저히 낮아짐
- rebase를 함으로써 깔끔한 커밋 기록을 가져갈 수 있음

판단은 @seokhwan-an , @somefood , @JoungMinJu 3명이서 해주시면 될 것 같아요 ㅎㅎ
